src/gui/widgets/ControllerView.cpp

```diff
@@ -23,176 +23,190 @@
+		
+	setModel( _model );
+}
 
-	hBox->addWidget(controlsButton);
 
-	m_nameLabel = new QLabel(_model->name(), this);
-	vBoxLayout->addWidget(m_nameLabel);
 
 
-	m_controllerDlg = getController()->createDialog( gui->mainWindow()->workspace() );
+ControllerView::~ControllerView()
+{
 
-	m_subWindow = gui->mainWindow()->addWindowedWidget( m_controllerDlg );
-	
-	Qt::WindowFlags flags = m_subWindow->windowFlags();
-	flags &= ~Qt::WindowMaximizeButtonHint;
-	m_subWindow->setWindowFlags( flags );
-	m_subWindow->setFixedSize( m_subWindow->size() );
+}
 
-	m_subWindow->setWindowIcon( m_controllerDlg->windowIcon() );
 
-	connect( m_controllerDlg, SIGNAL( closed() ),
-		this, SLOT( closeControls() ) );
 
-	m_subWindow->hide();
 
-	setWhatsThis( tr( "Controllers are able to automate the value of a knob, "
-				"slider, and other controls."  ) );
 
-	setModel( _model );
+void ControllerView::deleteController()
+{
+	emit( deleteController( this ) );
 }
 
 
 
 
-ControllerView::~ControllerView()
+void ControllerView::collapseController()
 {
-	if (m_subWindow)
+	if( m_controllerDlg->isHidden() )
 	{
-		delete m_subWindow;
+		m_controllerDlg->show();
+		setFixedHeight( m_controllerDlg->height() + 24 );
+		m_collapse->setIcon( embed::getIconPixmap( "stepper-down" ) );
+	}
+	else
+	{
+		m_collapse->setIcon( embed::getIconPixmap( "stepper-left" ) );
+		m_controllerDlg->hide();
+		setFixedHeight( 24 );
 	}
 }
 
 
 
 
-void ControllerView::editControls()
+void ControllerView::renameFinished()
 {
-	if( m_show )
-	{
-		m_subWindow->show();
-		m_subWindow->raise();
-		m_show = false;
-	}
-	else
+	m_nameLineEdit->setReadOnly( true );
+	Controller * c = castModel<Controller>();
+	QString new_name = m_nameLineEdit->text();
+	if( new_name != c->name() )
 	{
-		m_subWindow->hide();
-		m_show = true;
+		c->setName( new_name );
+		Engine::getSong()->setModified();
 	}
 }
 
 
 
 
-void ControllerView::closeControls()
+void ControllerView::rename()
 {
-	m_subWindow->hide();
-	m_show = true;
+	m_nameLineEdit->setReadOnly( false );
+	m_nameLineEdit->setFocus();
+	m_nameLineEdit->selectAll();
 }
 
 
-void ControllerView::deleteController()
-{
-	emit( deleteController( this ) );
-}
 
-void ControllerView::renameController()
-{
-	bool ok;
-	Controller * c = castModel<Controller>();
-	QString new_name = QInputDialog::getText( this,
-			tr( "Rename controller" ),
-			tr( "Enter the new name for this controller" ),
-			QLineEdit::Normal, c->name() , &ok );
-	if( ok && !new_name.isEmpty() )
-	{
-		c->setName( new_name );
-		m_nameLabel->setText( new_name );
-	}
-}
 
 
 void ControllerView::mouseDoubleClickEvent( QMouseEvent * event )
 {
-	renameController();
+	rename();
 }
 
 
 
+
 void ControllerView::modelChanged()
 {
 }
 
 
 
+
+void ControllerView::paintEvent(QPaintEvent* event)
+{
+	int titleBarHeight = 24;
+	QPainter p( this );
+	QRect rect( 1, 1, width()-2, titleBarHeight + 5 );
+	QLinearGradient grad( QPoint( 0, 0 ), QPoint( 0, 24 ) );
```

Comment:
This whole thing could be avoided if you just use the background property in css and define it in css. In the end you'd set the rectrangle fill to
```c++
p.fillRect( rect, painter.background() );
```